src/copy.c

```diff
@@ -406,54 +458,40 @@ TSCopyMultiInsertBufferCleanup(TSCopyMultiInsertInfo *miinfo, TSCopyMultiInsertB
 }
 
 /*
- * Write out all stored tuples in all buffers out to the tables.
- *
- * Once flushed we also trim the tracked buffers list down to size by removing
- * the buffers created earliest first.
- *
- * Callers should pass 'curr_rri' as the ResultRelInfo that's currently being
- * used.  When cleaning up old buffers we'll never remove the one for
- * 'curr_rri'.
+ * Write out all stored tuples of all buffers to the chunks. Also,
+ * cleanup the allocated buffers and free memory.
  */
 static inline void
-TSCopyMultiInsertInfoFlush(TSCopyMultiInsertInfo *miinfo, ResultRelInfo *curr_rri)
+TSCopyMultiInsertInfoFlushAndCleanup(TSCopyMultiInsertInfo *miinfo)
 {
-	ListCell *lc;
+	HASH_SEQ_STATUS status;
+	MultiInsertBufferEntry *entry;
 
-	foreach (lc, miinfo->multiInsertBuffers)
-	{
-		TSCopyMultiInsertBuffer *buffer = (TSCopyMultiInsertBuffer *) lfirst(lc);
+	hash_seq_init(&status, miinfo->multiInsertBuffers);
 
+	for (entry = hash_seq_search(&status); entry != NULL; entry = hash_seq_search(&status))
+	{
+		TSCopyMultiInsertBuffer *buffer = entry->buffer;
 		TSCopyMultiInsertBufferFlush(miinfo, buffer);
 
 		/*
 		 * Cleanup the buffer and finish the bulk insert. The chunk could
-		 * be closed (e.g., timescaledb.max_open_chunks_per_insert) later
+		 * be closed (e.g., due to timescaledb.max_open_chunks_per_insert)
 		 * and the bulk insert must be finished first.
 		 */
 		TSCopyMultiInsertBufferCleanup(miinfo, buffer);
 	}
 
-	list_free(miinfo->multiInsertBuffers);
-	miinfo->multiInsertBuffers = NIL;
+	/* All existing buffers are flushed and the multi-insert states
+	 * are freed. So, delete old hash map and create a new one for further
+	 * inserts.
+	 */
+	hash_destroy(miinfo->multiInsertBuffers);
```

Comment:
TSCopyMultiInsertInfoDestroy? TBH I'd delete that function, it's one line.

Reply:
I removed the function